// discord-html-transcripts/dist/static.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MINIFY_OPTIONS = exports.LanguageAliases = exports.DummyUser = exports.defaultPFP = void 0;
exports.defaultPFP = 'https://cdn.discordapp.com/embed/avatars/0.png';
exports.DummyUser = {
    bot: false,
    id: '00000000000',
    tag: "Unknown User#0000",
    name: "Unknown User",
    username: "Unknown User",
    hexAccentColor: "#FFFFFF",
    avatarURL: function () { return 'https://cdn.discordapp.com/embed/avatars/0.png'; }
};
exports.LanguageAliases = {
    'ts': 'typescript',
    'js': 'javascript',
    'py': 'python',
    'rb': 'ruby',
};
exports.default = {
    defaultPFP: exports.defaultPFP,
    DummyUser: exports.DummyUser,
    LanguageAliases: exports.LanguageAliases
};
exports.MINIFY_OPTIONS = {
    caseSensitive: false,
    collapseBooleanAttributes: true,
    collapseInlineTagWhitespace: true,
    collapseWhitespace: true,
    conservativeCollapse: true,
    keepClosingSlash: true,
    minifyCSS: true,
    minifyJS: true,
    minifyURLs: true,
    preserveLineBreaks: false,
    removeComments: true,
    removeAttributeQuotes: true,
    removeEmptyAttributes: true,
    removeOptionalTags: true,
    removeRedundantAttributes: true,
    removeScriptTypeAttributes: true,
    removeStyleLinkTypeAttributes: true,
    removeTagWhitespace: true,
    useShortDoctype: true
};
